pkg/storage/ingest/partition_offset_client.go

```diff
@@ -202,22 +196,24 @@ func (p *partitionOffsetClient) fetchPartitionOffset(ctx context.Context, partit
 	return listRes.Topics[0].Partitions[0].Offset, nil
 }
 
-// FetchTopicLastProducedOffsets fetches and returns the last produced offsets for all topic partitions. The offset is
-// -1 if a partition has been created but no record has been produced yet.
+// FetchPartitionsLastProducedOffsets fetches and returns the last produced offsets for all input partitions. The offset is
+// -1 if a partition has been created but no record has been produced yet. The returned offsets for each partition
+// are guaranteed to be always updated (no stale or cached offsets returned).
 //
 // The Kafka client used under the hood may retry a failed request until the retry timeout is hit.
-//
-// New partitions may not be immediately discovered by this function because the list of existing partitions is cached
-// for a short period of time under the hood. However, the returned offsets for the known partitions are guaranteed
-// to be always updated.
-func (p *partitionOffsetClient) FetchTopicLastProducedOffsets(ctx context.Context) (_ map[int32]int64, returnErr error) {
+func (p *PartitionOffsetClient) FetchPartitionsLastProducedOffsets(ctx context.Context, partitionIDs []int32) (_ map[int32]int64, returnErr error) {
+	// Skip lookup and don't track any metric if no partition was requested.
+	if len(partitionIDs) == 0 {
+		return nil, nil
```

Comment:
(nit) How about we return an empty map in this case?

```suggestion
		return make(map[int32]int64), nil
```

This is a redundant allocation, but it will preserve the invariant, that the resulting map can be accessed by key (i.e. no nil check is required if err is `nil`). The existing code already handles this. But this can make it more future-proof.

Reply:
you can index into a `nil` map: https://go.dev/play/p/5d7pZvb8i7g